### ISSUE TITLE
Memory leak in the RPM package scanning process

### ISSUE THREAD

**Issue by crolopez** [wazuh/wazuh#3584]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.9.2 | Syscollector | Manager and agent | Packages and sources | CentOS 7 |

A memory leak in this module has been found in the Valgrind report for another PR https://github.com/wazuh/wazuh/pull/3563#issuecomment-506238557. Specifically, it seems to affect the scanning of RPM packages.

These are the parts of the report that explain the memory leaks in Syscollector:

```
==31122== 128 bytes in 1 blocks are definitely lost in loss record 116 of 175
==31122==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==31122==    by 0x56CEA64: __os_malloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A040D: __env_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56CAC7F: __memp_init (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56CB589: __memp_open (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A42D3: __env_attach_regions (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A4574: __env_open (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5669A1E: __env_setup (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5688F5E: __db_open (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5682C1C: __db_open_pp (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x473560: sys_rpm_packages (syscollector_linux.c:431)
==31122==    by 0x47323F: sys_packages_linux (syscollector_linux.c:361)

==31122== 4,192 bytes in 1 blocks are definitely lost in loss record 168 of 175
==31122==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==31122==    by 0x56CEA64: __os_malloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A040D: __env_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C01CD: __memp_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C388E: __memp_fget (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5654A46: __ham_get_meta (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5655446: __ham_open (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56890CD: __db_open (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5682C1C: __db_open_pp (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x473560: sys_rpm_packages (syscollector_linux.c:431)
==31122==    by 0x47323F: sys_packages_linux (syscollector_linux.c:361)
==31122==    by 0x471862: wm_sys_main (syscollector_common.c:123)

==31122== 4,192 bytes in 1 blocks are definitely lost in loss record 169 of 175
==31122==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==31122==    by 0x56CEA64: __os_malloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A040D: __env_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C01CD: __memp_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C388E: __memp_fget (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x565A464: __ham_get_cpage (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x565A9FE: __ham_item_next (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x564F0F6: __hamc_get (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x566FFF8: __dbc_iget (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5680A2D: __dbc_get_pp (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x473D85: sys_rpm_packages (syscollector_linux.c:448)
==31122==    by 0x47323F: sys_packages_linux (syscollector_linux.c:361)

==31122== 255,712 bytes in 61 blocks are definitely lost in loss record 174 of 175
==31122==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==31122==    by 0x56CEA64: __os_malloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56A040D: __env_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C01CD: __memp_alloc (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x56C388E: __memp_fget (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x568A386: __db_goff (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5670451: __dbc_iget (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x5680A2D: __dbc_get_pp (in /var/ossec/lib/libwazuhext.so)
==31122==    by 0x473D85: sys_rpm_packages (syscollector_linux.c:448)
==31122==    by 0x47323F: sys_packages_linux (syscollector_linux.c:361)
==31122==    by 0x471862: wm_sys_main (syscollector_common.c:123)
==31122==    by 0x5A46DD4: start_thread (in /usr/lib64/libpthread-2.17.so)
```

In short, these calls have a memory pool that does not appear to be being released: 

https://github.com/wazuh/wazuh/blob/421c4fd9b5c142936b3f1922587c355076ae4e03/src/wazuh_modules/syscollector/syscollector_linux.c#L448

https://github.com/wazuh/wazuh/blob/421c4fd9b5c142936b3f1922587c355076ae4e03/src/wazuh_modules/syscollector/syscollector_linux.c#L431

For easy playback, launch multiple scans with the following configuration:

``` XML
  <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>5s</interval>
    <scan_on_start>yes</scan_on_start>
    <packages>yes</packages>
  </wodle>
```

The following error is a consequence of configuring small intervals and has nothing to do with the issue:

> 2019/06/27 07:44:29 wazuh-modulesd:syscollector: ERROR: Interval overtaken.

**Comment by chemamartinez** [comment]
Hi @crolopez,

I think the problem here is to stop the execution of `valgrind` while the scan is running.

The resources that cause that memory leaks are freed after the loop of reading the packages database, here: 
https://github.com/wazuh/wazuh/blob/bc2ba4c147b21b425d29f89851f120b9554ac949/src/wazuh_modules/syscollector/syscollector_linux.c#L591
https://github.com/wazuh/wazuh/blob/bc2ba4c147b21b425d29f89851f120b9554ac949/src/wazuh_modules/syscollector/syscollector_linux.c#L592

And no `return` statements are found in that loop. Here a couple of examples:

**Stopping the program during the package scan**

```
2019/06/27 06:28:54 wazuh-modulesd:syscollector[18537] syscollector_linux.c:351 at sys_packages_linux(): DEBUG: Starting installed packages inventory.
^C2019/06/27 06:28:55 wazuh-modulesd:syscollector[18537] syscollector_common.c:211 at wm_sys_cleanup(): INFO: Module finished.
==18537==
==18537== HEAP SUMMARY:
==18537==     in use at exit: 9,270,487 bytes in 242 blocks
==18537==   total heap usage: 8,863 allocs, 8,621 frees, 54,842,940 bytes allocated
==18537==
==18537== LEAK SUMMARY:
==18537==    definitely lost: 255,840 bytes in 62 blocks
==18537==    indirectly lost: 0 bytes in 0 blocks
==18537==      possibly lost: 19,482 bytes in 17 blocks
==18537==    still reachable: 8,995,165 bytes in 163 blocks
==18537==                       of which reachable via heuristic:
==18537==                         length64           : 223,848 bytes in 121 blocks
==18537==         suppressed: 0 bytes in 0 blocks
==18537== Rerun with --leak-check=full to see details of leaked memory
==18537==
==18537== For counts of detected and suppressed errors, rerun with: -v
==18537== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Waiting for the end of the package scan**

```
2019/06/27 06:30:02 wazuh-modulesd:syscollector[18543] syscollector_linux.c:351 at sys_packages_linux(): DEBUG: Starting installed packages inventory.
2019/06/27 06:30:08 wazuh-modulesd:syscollector[18543] syscollector_common.c:158 at wm_sys_main(): INFO: Evaluation finished.
^C2019/06/27 06:30:11 wazuh-modulesd:syscollector[18543] syscollector_common.c:211 at wm_sys_cleanup(): INFO: Module finished.
==18543==
==18543== HEAP SUMMARY:
==18543==     in use at exit: 370,543 bytes in 140 blocks
==18543==   total heap usage: 30,514 allocs, 30,374 frees, 204,198,563 bytes allocated
==18543==
==18543== LEAK SUMMARY:
==18543==    definitely lost: 0 bytes in 0 blocks
==18543==    indirectly lost: 0 bytes in 0 blocks
==18543==      possibly lost: 6,904 bytes in 6 blocks
==18543==    still reachable: 363,639 bytes in 134 blocks
==18543==                       of which reachable via heuristic:
==18543==                         length64           : 223,848 bytes in 121 blocks
==18543==         suppressed: 0 bytes in 0 blocks
==18543== Rerun with --leak-check=full to see details of leaked memory
==18543==
==18543== For counts of detected and suppressed errors, rerun with: -v
==18543== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Regards.

**Comment by crolopez** [comment]
I'm going to close this issue because it's a false positive, as @chemamartinez says.

Regards.
